cmake/Settings.cmake

```diff
@@ -0,0 +1,76 @@
+# This source file is part of the Swift.org open source project
+#
+# Copyright (c) 2024 Apple Inc. and the Swift project authors
+# Licensed under Apache License v2.0 with Runtime Library Exception
+#
+# See http://swift.org/LICENSE.txt for license information
+# See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+
+# Settings intended to be applied to every Swift target in this package.
+# Analogous to project-level build settings in an Xcode project.
+add_compile_options(
+  "SHELL:$<$<COMPILE_LANGUAGE:Swift>:-Xfrontend -require-explicit-sendable>")
+add_compile_options(
+  "SHELL:$<$<COMPILE_LANGUAGE:Swift>:-Xfrontend -enable-experimental-feature -Xfrontend StrictConcurrency>"
+  "SHELL:$<$<COMPILE_LANGUAGE:Swift>:-Xfrontend -enable-experimental-feature -Xfrontend AccessLevelOnImport>")
+add_compile_options(
+  "SHELL:$<$<COMPILE_LANGUAGE:Swift>:-Xfrontend -enable-upcoming-feature -Xfrontend ExistentialAny>"
+  "SHELL:$<$<COMPILE_LANGUAGE:Swift>:-Xfrontend -enable-upcoming-feature -Xfrontend InternalImportsByDefault>")
+
+# Settings which define commonly-used OS availability macros.
+#
+# These leverage a pseudo-experimental feature in the Swift compiler for
+# setting availability definitions, which was added in
+# [apple/swift#65218](https://github.com/apple/swift/pull/65218).
+add_compile_options(
+  "SHELL:$<$<COMPILE_LANGUAGE:Swift>:-Xfrontend -enable-experimental-feature -Xfrontend \"AvailabilityMacro=_mangledTypeNameAPI:macOS 11.0, iOS 14.0, watchOS 7.0, tvOS 14.0\">"
+  "SHELL:$<$<COMPILE_LANGUAGE:Swift>:-Xfrontend -enable-experimental-feature -Xfrontend \"AvailabilityMacro=_backtraceAsyncAPI:macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0\">"
+  "SHELL:$<$<COMPILE_LANGUAGE:Swift>:-Xfrontend -enable-experimental-feature -Xfrontend \"AvailabilityMacro=_clockAPI:macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0\">"
+  "SHELL:$<$<COMPILE_LANGUAGE:Swift>:-Xfrontend -enable-experimental-feature -Xfrontend \"AvailabilityMacro=_regexAPI:macOS 13.0, iOS 16.0, tvOS 16.0, watchOS 9.0\">"
+  "SHELL:$<$<COMPILE_LANGUAGE:Swift>:-Xfrontend -enable-experimental-feature -Xfrontend \"AvailabilityMacro=_swiftVersionAPI:macOS 13.0, iOS 16.0, tvOS 16.0, watchOS 9.0\">"
+  "SHELL:$<$<COMPILE_LANGUAGE:Swift>:-Xfrontend -enable-experimental-feature -Xfrontend \"AvailabilityMacro=_distantFuture:macOS 99.0, iOS 99.0, watchOS 99.0, tvOS 99.0\">")
+
+if(APPLE)
+  add_compile_definitions("SWT_TARGET_OS_APPLE")
+endif()
+if(WASI)
+  add_compile_definitions("SWT_NO_FILE_IO")
+endif()
+if(CMAKE_SYSTEM_NAME IN_LIST "iOS;watchOS;tvOS;visionOS;WASI")
+  add_compile_definitions("SWT_NO_EXIT_TESTS")
+endif()
+
+find_package(Git)
+if(Git_FOUND)
+  execute_process(
```

Comment:
I'm a tad concerned about this bit.
First, if you commit something that doesn't require CMake to re-run, the tag/commit won't get updated.
Second, it will always invalidate all of the files and we end up with a rebuild of _all the things_.
A custom-command that writes it to a file that exposes a "version function" (`getVersionString`) will ensure that incremental builds continue to work and only that one file gets rebuilt, and that it's run at build time, so it's kept up-to-date.

Trying to find a full example with the custom command and the fix. Most folks do accept the fact that re-running CMake will be required though. With that form, running `configure_file` on something like the snippet below (in a `Version.swift.in`) would be better.

```
public func getVersionString() {
    "@GIT_REVISION@"
}
```

That may need some adjustments to both builds though.

Reply:
We don't have a viable mechanism for emitting a Swift function like that when using SwiftPM, so we wouldn't be able to support it.

At least for my own nefarious purposes, I don't mind re-running CMake if the commit hash changes.